File: judge/src/tested/languages/c/templates/context.c
## Code to execute_module one context.

#include <stdio.h>

#include "values.h"
#include "${submission_name}.c"

## Import the language specific evaluators we will need.
% for name in evaluator_names:
    #include "${name}.c"
% endfor

static FILE* ${context_name}_value_file = NULL;
static FILE* ${context_name}_exception_file = NULL;

static void ${context_name}_write_separator() {
    fprintf(${context_name}_value_file, "--${secret_id}-- SEP");
    fprintf(${context_name}_exception_file, "--${secret_id}-- SEP");
    fprintf(stdout, "--${secret_id}-- SEP");
    fprintf(stderr, "--${secret_id}-- SEP");
}

## Send a value to TESTed.
## Use a macro, since there are no generics in C.
#undef send_value
#define send_value(value) write_value(${context_name}_value_file, value)

## Send the result of a language specific value evaluator to TESTed.
#undef send_specific_value
#define send_specific_value(r) send_evaluated(${context_name}_value_file, r)


int ${context_name}() {

    ${context_name}_value_file = fopen("${value_file}", "w");
    ${context_name}_exception_file = fopen("${exception_file}", "w");

    ${before}

    ${context_name}_write_separator();

    % if context_testcase.exists:
        char* args[] = {\
        % for argument in context_testcase.arguments:
            "${argument}", \
        % endfor
        };
        solution_main(${len(context_testcase.arguments)}, args);
    % endif

    ## Generate the actual tests based on the context.
    % for testcase in testcases:
        ${context_name}_write_separator();
        <%include file="statement.mako" args="statement=testcase.input_statement()" />;
        
    % endfor

    ${after}

    fclose(${context_name}_value_file);
    fclose(${context_name}_exception_file);
    return 0;
}

#ifndef INCLUDED
int main() {
    return ${context_name}();
}
#endif

File: judge/src/tested/languages/haskell/templates/context.hs
{-# LANGUAGE NamedFieldPuns #-}
## Code to execute one context.
<%! from tested.languages.generator import _TestcaseArguments %>
<%! from tested.serialisation import Statement, Expression, Assignment %>
<%! from tested.utils import get_args %>
module ${context_name} where

##################################
## Setup                        ##
##################################

import System.IO (hPutStr, stderr, stdout, hFlush)
import System.Environment
import qualified Values
import Control.Monad.Trans.Class
import Control.Exception
import EvaluationUtils

## Import the language specific evaluators we will need.
% for name in evaluator_names:
    import qualified ${name}
% endfor

## Import the solution.
import qualified ${submission_name}


value_file = "${value_file}"
exception_file = "${exception_file}"

## Write the separator and flush to ensure the output is in the files.
writeSeparator :: IO ()
writeSeparator = do
    hPutStr stderr "--${secret_id}-- SEP"
    hPutStr stdout "--${secret_id}-- SEP"
    appendFile value_file "--${secret_id}-- SEP"
    appendFile exception_file "--${secret_id}-- SEP"
    hFlush stdout
    hFlush stderr


##################################
## Predefined functions         ##
##################################

## Send a value to TESTed
sendValue :: Values.Typeable a => a -> IO ()
sendValue = Values.sendValue value_file

## Send an exception to TESTed
sendException :: Exception e => Maybe e -> IO ()
sendException = Values.sendException exception_file

## Send the result of a language specific value evaluator to TESTed.
sendSpecificValue :: EvaluationResult -> IO ()
sendSpecificValue r = Values.sendEvaluated value_file r

## Send the result of a language specific exception evaluator to TESTed.
sendSpecificException :: EvaluationResult -> IO ()
sendSpecificException r = Values.sendEvaluated exception_file r


## Exception handler
handleException :: Exception e => (Maybe e -> IO()) -> Either e a -> IO ()
handleException f (Left e) = f (Just e)
handleException f (Right _) = f (Nothing)

##################################
## Main testcase evaluators     ##
##################################

eEvaluateMain value = <%include file="statement.mako" args="statement=context_testcase.exception_function"/>

##################################
## Other testcase evaluators    ##
##################################
% for testcase in testcases:
    % if testcase.value_function:
        vEvaluate${loop.index} value = <%include file="statement.mako" args="statement=testcase.value_function"/>
    % endif

    eEvaluate${loop.index} value = <%include file="statement.mako" args="statement=testcase.exception_function"/>
% endfor

## Main function of the context.
main = do
    ${before}

    writeSeparator

    % if context_testcase.exists:
        let mainArgs = [\
            % for argument in context_testcase.arguments:
                <%include file="value.mako" args="value=argument"/>\
            % endfor
        ]
        result <- try (withArgs mainArgs ${submission_name}.main) :: IO (Either SomeException ())
        handleException eEvaluateMain result
    % endif

    % for testcase in testcases:
        writeSeparator

        ## In Haskell we do not actually have statements, so we need to keep them separate.
        ## Additionally, exceptions with "statements" are not supported at this time.
        % if isinstance(testcase.command, get_args(Assignment)):
            <%include file="statement.mako" args="statement=testcase.command,root=False" />
        % else:
            result${loop.index} <- catch
                (<%include file="statement.mako" args="statement=testcase.command,lifting=True" /> >>= \r -> vEvaluate${loop.index} r >> eEvaluate${loop.index} (Nothing :: Maybe SomeException))
                (\e -> eEvaluate${loop.index} (Just (e :: SomeException)))
        % endif

    % endfor

    putStr ""

    ${after}
